README.md
[![LinkedIn](https://raw.githubusercontent.com/thospfuller/awesome-backlinks/master/images/linkedin_32.png)](https://www.linkedin.com/in/thomasfuller/) [![Twitter](https://raw.githubusercontent.com/thospfuller/awesome-backlinks/master/images/twitter_32.png)](https://twitter.com/ThosPFuller) [![GitHub](https://raw.githubusercontent.com/thospfuller/awesome-backlinks/master/images/github_32.png)](https://github.com/thospfuller) [![Email](https://raw.githubusercontent.com/thospfuller/awesome-backlinks/master/images/email_32.png)](http://eepurl.com/b5jPPj) [![Coherent Logic Limited](https://github.com/thospfuller/awesome-backlinks/blob/master/images/CLSocialIconDarkBlue.png?raw=true)](https://coherentlogic.com?utm_source=rbitcoinchartsapi_on_gh)  [<img src="https://github.com/thospfuller/awesome-backlinks/blob/master/images/meetupcom_social_media_circled_network_64x64.png" height="32" width="32">](https://www.meetup.com/Washington-DC-CTO-Meetup-Group/)

# rbitcoinchartsapi: An R Package for the Bitcoincharts.com API

## Installing the rbitcoinchartsapi package from Drat

The following example demonstrates how to install the rbitcoinchartsapi from the [Drat repository located on GitHub](https://github.com/thospfuller/drat).

    install.packages(c("drat", "logging", "RJSONIO", "RCurl"))
    drat::addRepo("thospfuller")
    install.packages("rbitcoinchartsapi", type = "source")

Next we'll explore several examples that demonstrate how to use the various functions available in this package.

## Example 1: Introduction to the GetHistoricTradeData function

The GetHistoricTradeData function will return the 2000 most recent trades which are delayed by approximately 15 minutes.

The [symbols that are available can be found here](http://bitcoincharts.com/markets/).

Calling this function with invalid parameters will result in an empty data frame.

    library(rbitcoinchartsapi)
    params <- list (symbol="bitstampUSD")
    historicTradeData <- GetHistoricTradeData (params)
    with(historicTradeData, plot(unixtime, price))
    head(historicTradeData)

![The rbitcoinchartsapi GetHistoricTradeData function in action!](https://github.com/thospfuller/rbitcoinchartsapi/blob/master/images/rproject-rbitcoinchartsapi-package-gethistorictradedata-example.png?raw=true "The rbitcoinchartsapi GetHistoricTradeData function in action")

## Example 2: Introduction to the GetWeightedPrices function

The GetWeightedPrices function returns the weighted prices. 

[Bitcoincharts.com](http://www.bitcoincharts.com) offers weighted prices for several currencies that can be used, for example, to price goods and services in Bitcoins -- this will yield much lower fluctuations than using a single market's latest price.

Weighted prices are calculated for the last 24 hours, 7 days and 30 days; if there are no trades during an interval, such as no trade within 24 hours, then no value will be returned.

    weightedPrices <- GetWeightedPrices ()
    head(weightedPrices)

![The rbitcoinchartsapi GetWeightedPrices function in action!](https://github.com/thospfuller/rbitcoinchartsapi/blob/master/images/rproject-rbitcoinchartsapi-package-getweightedprices-example.png?raw=true "The rbitcoinchartsapi GetWeightedPrices function in action")

## Example 3: Introduction to the GetMarketData function

The GetMarketData function will return an array with elements for each market.

The [symbols that are available can be found here](http://bitcoincharts.com/markets/).

    params <- list (currency="USD")
    usd <- GetMarketData (params)
    head(usd)

![The rbitcoinchartsapi GetMarketData function in action!](https://github.com/thospfuller/rbitcoinchartsapi/blob/master/images/rproject-rbitcoinchartsapi-package-getmarketdata-example.png?raw=true "The rbitcoinchartsapi GetMarketData function in action")

NEWS
Package rbitcoinchartsapi
=========================
Version: 1.0.0 [2014-09-04]
o  Initial release


Version: 1.0.1 [2014-25-05]
o Documentation has been changed so that it is generated by roxygen2.


Version: 1.0.2 [2014-01-06]
o Examples have been improved.


Version: 1.0.3 [2014-08-06]
o One example has changed so as to prevent a possible null pointer exception.


Version: 1.0.4 [2014-14-06]
o Fixed a minor issue with several tryCatch statements.


Version: 1.0.5 [2022-16-05]
o Minor improvements and documentation updates.

R/rbitcoinchartsapi.R
#' 
#' @title R Bitcoin Charts API Package
#' 
#' @details This package provides functions for acquiring data directly from \href{http://www.bitcoincharts.com}{Bitcoincharts}. 
#'
#' @description This package provides functions for acquiring data directly from \href{http://www.bitcoincharts.com}{Bitcoincharts}.
#'
#' @import RJSONIO RCurl logging
#' @importFrom utils read.csv
#'
#' @name rbitcoinchartsapi
#'
#' @docType package
#'
NULL

#' This function returns the weighted prices. 
#'
#' \href{http://www.bitcoincharts.com}{Bitcoincharts} offers weighted prices
#' for several currencies that can be used, for example, to price goods and
#' services in Bitcoins -- this will yield much lower fluctuations than using
#' a single market's latest price.
#'
#' @return Weighted prices are calculated for the last 24 hours, 7 days and 30
#'  days; if there are no trades during an interval, such as no trade within 24
#'  hours, then no value will be returned.
#'
#' @examples
#'  tryCatch(
#'      weightedPrices <- GetWeightedPrices (),
#'      error =
#'          function (e) {    
#'              print (
#'                  paste (
#'                      "An exception was thrown -- details follow: ",
#'                      e,
#'                      sep=""
#'                  )
#'              )
#'          }
#'      )
#'
#' @export
#'
GetWeightedPrices <- function () {
  
    logging::logdebug("GetWeightedPrices: function begins.")
  
    data <- getURL("http://api.bitcoincharts.com/v1/weighted_prices.json")
    dataFrame <- RJSONIO::fromJSON(data)
    
    logging::logdebug("GetWeightedPrices: function ends.")
    
    return (dataFrame)
}

#' This function will return an array with elements for each market.
#'
#' General market data can be accessed \href{http://api.bitcoincharts.com/v1/markets.json}{here}.
#'
#' @param params Any parameter accepted by this web service call -- see \href{http://bitcoincharts.com/about/markets-api/}{here}
#'
#' @examples
#'  params <- list (currency="USD")
#'  tryCatch(
#'      usd <- GetMarketData (params),
#'      error =
#'          function (e) {    
#'              print (
#'                  paste (
#'                      "An exception was thrown -- details follow: ",
#'                      e,
#'                      sep=""
#'                  )
#'              )
#'          }
#'      )
#'
#' @export
#'
GetMarketData <- function (params) {
  
    logging::logdebug("GetMarketData: function begins.")
  
    data <- getForm("http://api.bitcoincharts.com/v1/markets.json", .params=params)
    dataFrame <- RJSONIO::fromJSON(data)
    
    logging::logdebug("GetMarketData: function ends.")
    
    return (dataFrame)
}

#' This function will return the 2000 most recent trades which are delayed by
#' approximately 15 minutes. 
#'
#' The symbols that are available can be found \href{http://bitcoincharts.com/markets/}{here}.
#'
#' Note that calling this function with invalid parameters will result in an
#' empty data frame.
#'
#' @param params Any parameter accepted by this web service call -- see
#' \href{http://bitcoincharts.com/about/markets-api/}{here}.
#'
#' @examples
#'  params <- list (symbol="bitstampUSD")
#'  tryCatch(
#'      historicTradeData <- GetHistoricTradeData (params),
#'      error =
#'          function (e) {    
#'              print (
#'                  paste (
#'                      "An exception was thrown -- details follow: ",
#'                      e,
#'                      sep=""
#'                  )
#'              )
#'          }
#'      )
#'  with(historicTradeData, plot(unixtime, price))
#'
#' @export
#'
GetHistoricTradeData <- function (params) {

    logging::logdebug("GetHistoricTradeData: function begins.")
  
    #
    # http://api.bitcoincharts.com/v1/trades.csv?symbol=bitstampUSD
    #
    # params <- list (symbol="krakenEUR", start=1303100000)
    # historicTradeData <- GetHistoricTradeData (params)
    #

    data <- getForm("http://api.bitcoincharts.com/v1/trades.csv", .params=params)

    tempCsvFile <- tempfile(pattern = "historicTradeData", tmpdir = tempdir(), fileext = ".csv")

    writtenResult <- write (data, file=tempCsvFile, append=FALSE)

    results <- utils::read.csv (tempCsvFile, header = FALSE)

    unlink (tempCsvFile)

    colnames (results) <- c("unixtime", "price", "amount")

    logging::logdebug("GetHistoricTradeData: function ends")
    
    return (results)
}
